### PR TITLE
[1.20.5] Remove uses of `Event.Result` and update events to have their own Result classes

### PR DESCRIPTION
**Pull request by Shadows-of-Fire**
## What
This PR is a work-in-progress for removing all uses of `Event.Result` from existing events.  Changes to an individual event will be grouped into a single commit until the first pass is completed.

## Why
Uses of the abstracted `Event.Result` enum induces the following readability issues:
1. Result Naming - Currently, all results are named `ALLOW` / `DEFAULT` / `DENY` for any result-based event, even when this is not a very descriptive name.
2. Result Documentation - The effects of an event result should be located in the definition of that enum value. Currently, this documentation lives in the class body.
3. Result-Based Nature - It is not obvious to a reader what the impact of the `@HasResult` annotation are, and it is unexpected that the `getResult` / `setResult` methods exist on all children of `Event`.
  a. This is the same reason why `@Cancellable` was removed in https://github.com/neoforged/Bus/pull/20

There are also technical challenges with continuing to support the annotation-based model, since it means we must run a transformer to inject real implementations of `getResult` / `setResult` based on the presence of the annotation.

## Progress
The following events need to be updated:
- [x] `MouseButtonPressed.Post`
- [x] `MouseButtonReleased.Post`
- [x] ~~`LivingPackSizeEvent`~~ renamed to `SpawnClusterSizeEvent`
- [x] `MobEffectEvent.Applicable` (Fixed a bug where this wouldn't fire for withers)
- [x] `MobSpawnEvent.SpawnPlacementCheck`
- [x] `MobSpawnEvent.PositionCheck` (Fixed a bug where the `BaseSpawner` wasn't getting passed through when applicable)
- [x] ~~`MobSpawnEvent.AllowDespawn`~~ renamed to `MobDespawnEvent`
- [x] ~~`ZombieEvent.SummonAidEvent`~~ removed. Supplanted by other events.
- [x] `CriticalHitEvent`
- [x] ~~`SleepingLocationCheckEvent`~~ (This and the below event have been fused to `CanContinueSleepingEvent`)
- [x] ~~`SleepingTimeCheckEvent`~~
- [x] `EntityMobGriefingEvent`
- [x] `BlockEvent.CreateFluidSourceEvent`
- [x] `CropGrowEvent.Pre`
- [x] ~~`SaplingGrowTreeEvent`~~ renamed to `BlockGrowFeatureEvent `

## Notes
Depends on https://github.com/neoforged/Bus/pull/30 - this PR will not build locally or in CI unless using a version of Bus with this PR merged.

### CODE REVIEW

**Comment by Shadows-of-Fire** [comment]
Still 92 compile errors.  My original list is missing a few entries.

**Comment by Shadows-of-Fire** [comment]
Remaining events:

- [x] ~~FillBucketEvent~~ Deleted. This is a very old event (circa 1.2.5) that is fully superseded by `PlayerInteractEvent.RightClickItem` with an `item == Items.BUCKET` check. It also has been erroneously firing for non-empty buckets for many MC versions (since at least 1.16)
- [x] PlayerSpawnPhantomsEvent
- [x] RenderNameTagEvent

File: patches/net/minecraft/world/entity/LivingEntity.java.patch [hunk 1]
```diff
@@ -575,9 +595,9 @@
  
      private boolean checkBedExists() {
 -        return this.getSleepingPos().map(p_337701_ -> this.level().getBlockState(p_337701_).getBlock() instanceof BedBlock).orElse(false);
-+        return this.getSleepingPos().map((p_289310_) -> {
-+            return net.neoforged.neoforge.event.EventHooks.fireSleepingLocationCheck(this, p_289310_);
-+        }).orElse(false);
++        // Neo: Overwrite the vanilla instanceof BedBlock check with isBed and fire the sleeping location check event.
```
Comment (Shadows-of-Fire) [inline patches/net/minecraft/world/entity/LivingEntity.java.patch]: Comment is slightly stale here since this event was renamed

File: src/main/java/net/neoforged/neoforge/event/entity/player/CriticalHitEvent.java [hunk 2]
```diff
@@ -7,72 +7,96 @@
...
+    }
+
+    /**
+     * The damage multiplier is applied to the base attack's damage if the attack {@linkplain #isCriticalHit() critically hits}.
+     * <p>
+     * A damage multiplier of 1.0 will not change the damage, a value of 1.5 will increase the damage by 50%, and so on.
+     * 
+     * @param modifier The new damage modifier.
+     */
+    public float getDmgMultiplier() {
```
Comment (Matyrobbrt) [inline src/main/java/net/neoforged/neoforge/event/entity/player/CriticalHitEvent.java]: Please avoid using abbreviations in method names, and change this to `getDamageMultiplier`.

File: src/main/java/net/neoforged/neoforge/event/level/BlockGrowFeatureEvent.java [hunk 3]
```diff
@@ -65,16 +58,30 @@ public BlockPos getPos() {
     }
 
     /**
+     * Changes the feature that will be grown. May not set
+     * 
      * @param feature a {@linkplain Holder} referencing a tree feature to be placed instead of the current feature.
      */
-    public void setFeature(@Nullable Holder<ConfiguredFeature<?, ?>> feature) {
+    public void setFeature(Holder<ConfiguredFeature<?, ?>> feature) {
```
Comment (Matyrobbrt) [inline src/main/java/net/neoforged/neoforge/event/level/BlockGrowFeatureEvent.java]: Why did you remove the annotation?

File: src/main/java/net/neoforged/neoforge/event/entity/living/SpawnClusterSizeEvent.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+     * To see the default size, use {@link Mob#getMaxSpawnClusterSize()}
+     * 
+     * @return The max spawn cluster size
+     */
+    public int getSize() {
+        return size;
+    }
+
+    /**
+     * Changes the max cluster size for the entity.
```
Comment (Matyrobbrt) [inline src/main/java/net/neoforged/neoforge/event/entity/living/SpawnClusterSizeEvent.java:35]: This should document whether cancelling the event is required too. (and if so, cancel automatically)

**Review by Matyrobbrt — CHANGES_REQUESTED** [review]

File: src/main/java/net/neoforged/neoforge/event/level/BlockGrowFeatureEvent.java (hunk 3 above)
Comment (Shadows-of-Fire) [reply src/main/java/net/neoforged/neoforge/event/level/BlockGrowFeatureEvent.java]: Think I was in the middle of messing with this and got distracted (as indicated by the unfinished comment in the same class)

File: src/main/java/net/neoforged/neoforge/event/entity/living/SpawnClusterSizeEvent.java (hunk 4 above)
Comment (Shadows-of-Fire) [reply src/main/java/net/neoforged/neoforge/event/entity/living/SpawnClusterSizeEvent.java:35]: This is not a cancellable event

**Review by Technici4n — APPROVED** [review]

**Review by Matyrobbrt — APPROVED** [review]
